src/ImageRunModal.jsx

```diff
@@ -27,13 +28,107 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        this.setState(state => {
+            let ports = state.publish.concat();
+            ports.splice(idx, 1);
+            if (ports.length === 0)
+                ports.push({ containerPort: null, hostPort: null, protocol: 'TCP' });
+            return { publish: ports };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    addPort() {
+        this.setState(state => {
+            return { publish: [...state.publish, { containerPort: null, hostPort: null, protocol: 'TCP' }] };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    onPortChange(idx, field, value) {
+        this.setState(state => {
+            let ports = state.publish.concat();
+            ports[idx][field] = value || null;
+            return { publish: ports };
+        }, () => this.props.onChange(this.state.publish.concat()));
+    }
+
+    render () {
+        const { id } = this.props;
+        const dialogValues = this.state;
+        return (
+            <React.Fragment>
+                {
+                    dialogValues.publish.map((port, idx) =>
+                        (
+                            <div className='publish-port-form ct-form-layout' key={'published-port-' + idx}>
```

Comment:
Using index as key [goes against its purpose](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). You need to give an intrinsic property of what you render, not an external index. I suggest using the container port (possibly concatenated with the protocol, see above), that's guaranteed to be unique and also intrinsic. Also, you don't need the string prefix, the key must be unique only within the enclosing `{ }` as far as I know (@KKoukiou, is that right?)

Reply:
Can I use the index together with the ports? Because a user might click the  plus button several times, and then you'll have multiple rows with the same key.

Reply:
Also using a key based on the values of the form causes an unfocus each time you enter a single number.

Reply:
Right, I didn't consider multiple empty inputs. So host port + index sounds fine; it will make the non-empty entries a  bit more ambiguous, but at worst we'll get a few extra renders there.